### PR TITLE
*: add xtrabackup manger

### LINKED ISSUE #116: [feature]add xtrabackup manger
**Is your feature request related to a problem? Please describe.**

we need xtrabackup to cloud for the cluster
**Describe the solution you'd like**
Make a backup like this:
```shell
kubectl apply -f mysql_v1alpha1_backup.yaml 
```
To backup the database data to S3 storage , or Persisten volume
Then could use the backup copy to create cluster , add RestoreFrom feild in mysql_v1alpha1_cluster.yaml
```shell
kubectl apply -f mysql_v1alpha1_cluster.yaml 
```
create the cluster from backup copy

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by acekingke**
### Which issue(s) this PR fixes?

Fixes #116

### What this PR does?

Summary:
1.  backup to S3 and restore cluster from S3
2. backup to NFS and restore cluster from NFS
3. Clone from leader by using backup and restore

### CODE REVIEW

**Comment by andyli029** [comment]
Fix the commit log

File: DEPLOY.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
+# mysql-operator
```
Comment (zhyass) [inline DEPLOY.md:1]: move to the docs dir

File: Dockerfile.sidecar [hunk 2]
```diff
@@ -2,12 +2,14 @@
 #  Build Sidecar
 ###############################################################################
 # Build the manager binary
-FROM golang:1.16 as builder
+FROM golang:1.15 as builder
```
Comment (zhyass) [inline Dockerfile.sidecar:5]: use 1.16

File: Dockerfile.sidecar [hunk 3]
```diff
@@ -19,17 +21,39 @@ COPY sidecar/ sidecar/
 COPY utils/ utils/
 
 # Build
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o bin/sidecar cmd/sidecar/main.go
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o bin/sidecar cmd/sidecar/main.go
```
Comment (zhyass) [inline Dockerfile.sidecar:24]: delete `GO111MODULE=on`

File: Dockerfile.sidecar [hunk 4]
```diff
@@ -19,17 +21,39 @@ COPY sidecar/ sidecar/
...
+    apt-get update; \
+    apt-get install -y --no-install-recommends gnupg2 wget lsb-release curl; \
+    wget -P /tmp --no-check-certificate https://repo.percona.com/apt/percona-release_latest.$(lsb_release -sc)_all.deb; \
+    dpkg -i /tmp/percona-release_latest.$(lsb_release -sc)_all.deb; \
+    apt-get update; \
+    apt-get install -y --no-install-recommends ${XTRABACKUP_PKG}
 WORKDIR /
 COPY --from=builder /workspace/bin/sidecar /usr/local/bin/sidecar
-
+COPY ./hack/restore.sh /restore.sh
```
Comment (zhyass) [inline Dockerfile.sidecar:55]: COPY --chown=1001:1001 ./hack/restore.sh /restore.sh

You can set `./hack/restore.sh` 's permissions locally, then push the update

File: api/v1alpha1/mysqlbackup_types.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2021 zhyass.
```
Comment (zhyass) [inline api/v1alpha1/mysqlbackup_types.go:2]: Copyright 2021 RadonDB

File: api/v1alpha1/mysqlbackup_types.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// MysqlBackupSpec defines the desired state of MysqlBackup
+type MysqlBackupSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Foo is an example field of MysqlBackup. Edit mysqlbackup_types.go to remove/update
```
Comment (zhyass) [inline api/v1alpha1/mysqlbackup_types.go:32]: delete the comment

File: PROJECT [hunk 7]
```diff
@@ -18,4 +18,13 @@ resources:
   group: mysql
   kind: Status
   version: v1alpha1
+- api:
+    crdVersion: v1
+    namespaced: true
+  controller: true
+  domain: radondb.io
+  group: mysql
+  kind: MysqlBackup
```
Comment (zhyass) [inline PROJECT:27]: Maybe use `Backup` much better?

File: PROJECT [hunk 8]
```diff
@@ -18,4 +18,13 @@ resources:
...
+  path: github.com/radondb/radondb-mysql-kubernetes/api/v1alpha1
+  version: v1
```
Comment (zhyass) [inline PROJECT:29]: use v1alpha1 pls

**Review by andyli029 — COMMENTED** [review]
Add #116 to the tail of commit log

File: charts/mysql-operator/values.yaml [hunk 9]
```diff
@@ -14,8 +14,8 @@ nameOverride: ""
 fullnameOverride: ""
 
 manager:
-  image: zhyass/manager
-  tag: 0.1
+  image: acekingke/controller
```
Comment (zhyass) [inline charts/mysql-operator/values.yaml:17]: Don't modify temporarily, we'll build and push the image to xenondb or radondb org in docker hub.

File: cluster/container/backup.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+			Name:  "NAMESPACE",
+			Value: c.Namespace,
+		},
+		{
+			Name:  "SERVICE_NAME",
+			Value: c.GetNameForResource(utils.HeadlessSVC),
+		},
+	}
+	if len(sctName) != 0 {
+		envs = append(envs, []corev1.EnvVar{
```
Comment (zhyass) [inline cluster/container/backup.go:44]: envs = append(envs,
    getEnvVarFromSecret(sctName, "S3_ENDPOINT", "s3-endpoint", false),
    getEnvVarFromSecret(sctName, "S3_ENDPOINT", "s3-endpoint", false),
    getEnvVarFromSecret(sctName, "S3_ACCESSKEY", "s3-access-key", true),
    getEnvVarFromSecret(sctName, "S3_SECRETKEY", "s3-secret-key", true),
    getEnvVarFromSecret(sctName, "S3_BUCKET", "s3-bucket", true),
)

File: cluster/container/backup.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+func (c *backupSidecar) getPorts() []corev1.ContainerPort {
+	return []corev1.ContainerPort{
+		{
+			Name:          utils.XBackupPortName,
+			ContainerPort: utils.XBackupPort,
+		},
+	}
+}
+
+func (c *backupSidecar) getLivenessProbe() *corev1.Probe {
```
Comment (zhyass) [inline cluster/container/backup.go:72]: Maybe need livenessprobe, you can set ad readinessprobe

File: cluster/container/init_sidecar.go [hunk 12]
```diff
@@ -89,7 +99,20 @@ func (c *initSidecar) getEnvVars() []corev1.EnvVar {
 		getEnvVarFromSecret(sctName, "OPERATOR_USER", "operator-user", true),
 		getEnvVarFromSecret(sctName, "OPERATOR_PASSWORD", "operator-password", true),
 	}
-
+	if len(c.Spec.BackupSecretName) != 0 {
+		envs = append(envs, []corev1.EnvVar{
```
Comment (zhyass) [inline cluster/container/init_sidecar.go:103]: The same question

File: cmd/manager/main.go [hunk 13]
```diff
@@ -94,6 +94,15 @@ func main() {
 		setupLog.Error(err, "unable to create controller", "controller", "Status")
 		os.Exit(1)
 	}
+	if err = (&controllers.MysqlBackupReconciler{
+		Client:   mgr.GetClient(),
+		Scheme:   mgr.GetScheme(),
+		Recorder: mgr.GetEventRecorderFor("controller.mysqlBackup"),
+		//Opt:      options.GetOptions(),
```
Comment (zhyass) [inline cmd/manager/main.go:101]: remove the unused code

File: config/manager/kustomization.yaml [hunk 14]
```diff
@@ -5,6 +5,12 @@ generatorOptions:
   disableNameSuffixHash: true
 
 configMapGenerator:
-- name: manager-config
-  files:
+- files:
   - controller_manager_config.yaml
+  name: manager-config
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (zhyass) [inline config/manager/kustomization.yaml:11]: The code generated by make deploy, maybe remove is better.

File: config/samples/backup_secret.yaml [hunk 15]
```diff
@@ -0,0 +1,11 @@
+kind: Secret
+apiVersion: v1
+metadata:
+  name: sample-backup-secret
+  namespace: default
+data:
+  s3-endpoint: aHR0cDovL3MzLnNoMWEucWluZ3N0b3IuY29t
```
Comment (zhyass) [inline config/samples/backup_secret.yaml:7]: maybe remove the value?

File: config/samples/mysql_v1alpha1_cluster.yaml [hunk 16]
```diff
@@ -53,8 +55,8 @@ spec:
         memory: 128Mi
 
   podSpec:
-    imagePullPolicy: IfNotPresent
-    sidecarImage: zhyass/sidecar:0.1
+    imagePullPolicy: Always
```
Comment (zhyass) [inline config/samples/mysql_v1alpha1_cluster.yaml:58]: don't modify, we'll push the image to xenondb or radondb

File: config/samples/mysql_v1alpha1_cluster.yaml [hunk 17]
```diff
@@ -77,4 +79,4 @@ spec:
     accessModes:
     - ReadWriteOnce
     #storageClass: ""
-    size: 10Gi
+    size: 30Gi
```
Comment (zhyass) [inline config/samples/mysql_v1alpha1_cluster.yaml:82]: The default is 10Gi, dont modify

File: config/samples/nfs_rc.yaml [hunk 18]
```diff
@@ -0,0 +1,34 @@
+
+apiVersion: v1
```
Comment (zhyass) [inline config/samples/nfs_rc.yaml:2]: Whether need the file

File: config/samples/nfs_server.yaml [hunk 19]
```diff
@@ -0,0 +1,14 @@
+kind: Service
```
Comment (zhyass) [inline config/samples/nfs_server.yaml:1]: Whether need the file, maybe removed or you can add docs to describe

File: controllers/mysqlbackup_controller.go [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"github.com/presslabs/controller-util/syncer"
+	apiv1alpha1 "github.com/radondb/radondb-mysql-kubernetes/api/v1alpha1"
```
Comment (zhyass) [inline controllers/mysqlbackup_controller.go:26]:
```
import (
	"fmt"
	"reflect"

	"github.com/presslabs/controller-util/syncer"

         apiv1alpha1 "github.com/radondb/radondb-mysql-kubernetes/api/v1alpha1"
)
```

The local pkg write at the end.

File: controllers/mysqlbackup_controller.go [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	// Set defaults on backup
+	r.Scheme.Default(backup.Unwrap())
+
+	// save the backup for later check for diff
+	savedBackup := backup.Unwrap().DeepCopy()
+	log.Info(fmt.Sprintf("saveBackup %v", savedBackup))
+	// cluster name should be specified for a backup
```
Comment (zhyass) [inline controllers/mysqlbackup_controller.go:84]: remove the unused code

File: controllers/mysqlbackup_controller.go [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MysqlBackupReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apiv1alpha1.MysqlBackup{}).
+		Complete(r)
+}
+func (r *MysqlBackupReconciler) updateBackup(savedBackup *apiv1alpha1.MysqlBackup, backup *mysqlbackup.MysqlBackup) error {
```
Comment (zhyass) [inline controllers/mysqlbackup_controller.go:109]: format the code
```
}

func
```

File: controllers/mysqlbackup_controller.go [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
```
Comment (zhyass) [inline controllers/mysqlbackup_controller.go:24]: pls use `"sigs.k8s.io/controller-runtime/pkg/log"`

File: api/v1alpha1/mysqlbackup_types.go (hunk 5 above)
Comment (andyli029) [inline api/v1alpha1/mysqlbackup_types.go:2]: Fix the file head

File: hack/backup_local.sh [hunk 24]
```diff
@@ -0,0 +1,40 @@
+ if [ $# != 1 ] ; then
```
Comment (zhyass) [inline hack/backup_local.sh:1]: #!/bin/sh or #!/bin/bash in the head

File: mysqlbackup/mysqlbackup.go [hunk 25]
```diff
@@ -0,0 +1,41 @@
+package mysqlbackup
+
+import (
+	"fmt"
+
+	v1alhpa1 "github.com/radondb/radondb-mysql-kubernetes/api/v1alpha1"
```
Comment (zhyass) [inline mysqlbackup/mysqlbackup.go:6]: The same question

File: mysqlbackup/syncer/job.go [hunk 26]
```diff
@@ -0,0 +1,157 @@
+package syncer
+
+import (
+	"fmt"
+
+	"github.com/presslabs/controller-util/syncer"
+	apiv1 "github.com/radondb/radondb-mysql-kubernetes/api/v1alpha1"
```
Comment (zhyass) [inline mysqlbackup/syncer/job.go:7]: apiv1alpha1

File: sidecar/init.go [hunk 27]
```diff
@@ -36,16 +38,83 @@ func NewInitCommand(cfg *Config) *cobra.Command {
 		Use:   "init",
 		Short: "do some initialization operations.",
 		Run: func(cmd *cobra.Command, args []string) {
-			if err := runInitCommand(cfg); err != nil {
-				log.Error(err, "init command failed")
-				os.Exit(1)
+			if err := runCloneAndInit(cfg); err != nil {
```
Comment (zhyass) [inline sidecar/init.go:41]: Why run init when clone and init error.

**Review by zhyass — CHANGES_REQUESTED** [review]

File: Dockerfile.sidecar [hunk 28]
```diff
@@ -19,17 +21,39 @@ COPY sidecar/ sidecar/
...
 
 # Build
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o bin/sidecar cmd/sidecar/main.go
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o bin/sidecar cmd/sidecar/main.go
 
 ###############################################################################
 #  Docker image for Sidecar
 ###############################################################################
-FROM alpine:3.13
+FROM ubuntu:focal
```
Comment (andyli029) [inline Dockerfile.sidecar:29]: The reason is?
Need discuss to @zhyass

File: charts/mysql-operator/values.yaml [hunk 29]
```diff
@@ -14,8 +14,8 @@ nameOverride: ""
...
+  tag: v01
```
Comment (andyli029) [inline charts/mysql-operator/values.yaml:18]: The same question

File: Dockerfile.sidecar [hunk 30]
```diff
@@ -2,12 +2,14 @@
...
 # Build the manager binary
-FROM golang:1.16 as builder
+FROM golang:1.15 as builder
 
 WORKDIR /workspace
 # Copy the Go Modules manifests
 COPY go.mod go.mod
 COPY go.sum go.sum
+
+
```
Comment (andyli029) [inline Dockerfile.sidecar:12]: Delete the line

File: cluster/container/backup.go [hunk 31]
```diff
@@ -0,0 +1,108 @@
+package container
```
Comment (andyli029) [inline cluster/container/backup.go:1]: Add the file head

File: cluster/container/backup.go [hunk 32]
```diff
@@ -0,0 +1,108 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/radondb/radondb-mysql-kubernetes/cluster"
+	"github.com/radondb/radondb-mysql-kubernetes/utils"
+)
+
+type backupSidecar struct {
+	*cluster.Cluster
+
```
Comment (andyli029) [inline cluster/container/backup.go:13]: Delete the space

File: cmd/sidecar/main.go [hunk 33]
```diff
@@ -51,10 +52,61 @@ func main() {
...
-
+	stop := make(chan struct{})
 	initCmd := sidecar.NewInitCommand(cfg)
 	cmd.AddCommand(initCmd)
+	takeBackupCmd := &cobra.Command{
+		Use:   "backup",
+		Short: "Take a backup from node and push it to rclone path.",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return fmt.Errorf("require two arguments. source host and destination bucket")
```
Comment (andyli029) [inline cmd/sidecar/main.go:63]: `len(args) != 1`
require two arguments?

File: Dockerfile.sidecar (hunk 28 above)
Comment (acekingke) [reply Dockerfile.sidecar:29]: alpine cannot install xtrabackup

File: config/samples/mysql_v1alpha1_cluster.yaml (hunk 17 above)
Comment (andyli029) [reply config/samples/mysql_v1alpha1_cluster.yaml:82]: How about to Change to 20Gi? KDE is limited to 20Gi

File: controllers/mysqlbackup_controller.go [hunk 34]
```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2021 zhyass.
```
Comment (andyli029) [inline controllers/mysqlbackup_controller.go:2]: Fix

File: config/samples/mysql_v1alpha1_cluster.yaml [hunk 35]
```diff
@@ -53,8 +55,8 @@ spec:
         memory: 128Mi
 
   podSpec:
-    imagePullPolicy: IfNotPresent
-    sidecarImage: zhyass/sidecar:0.1
+    imagePullPolicy: Always
+    sidecarImage: acekingke/sidecar:v01
```
[the remark this answers is not in the document]
Comment (runkecheng) [reply config/samples/mysql_v1alpha1_cluster.yaml:59]: ack

File: hack/backup_local.sh [hunk 36]
```diff
@@ -0,0 +1,40 @@
...
+	until $ctrl get pod/backup-access -o jsonpath='{.status.containerStatuses[0].ready}' 2>/dev/null | grep -q 'true'; do
+		sleep 1
+		echo -n .
+	done
+	echo "[done]"
+    # kubectl cp backup-access:backup /root/backup
+
+# curl --user sys_backups:sys_backups sample-mysql-0.sample-mysql.default:8082/xbackup
+
+# curl --user sys_backups:sys_backups sample-mysql-0.sample-mysql.default:8082/download|xbstream -x -C /backup
```
Comment (andyli029) [inline hack/backup_local.sh:40]: Return

File: sidecar/init.go [hunk 37]
```diff
@@ -36,16 +38,83 @@ func NewInitCommand(cfg *Config) *cobra.Command {
...
 			}
+
 		},
 	}
 
 	return cmd
 }
 
+func CheckServiceExist(cfg *Config, service string) bool {
+	serviceURl := fmt.Sprintf("http://%s-%s:%v%s", cfg.ClusterName, service, utils.XBackupPort, "/health")
```
Comment (andyli029) [inline sidecar/init.go:56]: serviceURl
->
serviceURL?

File: sidecar/server.go [hunk 38]
```diff
@@ -0,0 +1,221 @@
+package sidecar
```
Comment (andyli029) [inline sidecar/server.go:1]: Add the file head

File: hack/restore.sh [hunk 39]
```diff
@@ -0,0 +1,46 @@
...
+--parallel=10 $RESTORE_FROM \
+--insecure |xbstream -xv -C /root/backup
+# prepare redolog
+xtrabackup --defaults-file=/etc/mysql/my.cnf --use-memory=3072M --prepare --apply-log-only --target-dir=/root/backup
+# prepare data
+xtrabackup --defaults-file=/etc/mysql/my.cnf --use-memory=3072M --prepare --target-dir=/root/backup
+chown -R mysql.mysql /root/backup
+xtrabackup --defaults-file=/etc/mysql/my.cnf --datadir=/var/lib/mysql --copy-back --target-dir=/root/backup
+chown -R mysql.mysql /var/lib/mysql
+rm -rf /root/backup
```
Comment (andyli029) [inline hack/restore.sh:46]: Fix

File: mysqlbackup/status.go [hunk 40]
```diff
@@ -0,0 +1,72 @@
+package mysqlbackup
```
Comment (andyli029) [inline mysqlbackup/status.go:1]: Add the file head

File: utils/constants.go [hunk 41]
```diff
@@ -83,6 +90,9 @@ const (
 	ScriptsVolumeMountPath  = "/scripts"
 	XenonVolumeMountPath    = "/etc/xenon"
 	InitFileVolumeMountPath = "/docker-entrypoint-initdb.d"
+	SideCarImage            = "acekingke/sidecar:v01"
```
Comment (andyli029) [inline utils/constants.go:93]: Fix

File: DEPLOY.md [hunk 42]
```diff
@@ -0,0 +1,106 @@
...
+Uninstall the operator name `test`:
+
+```shell
+helm uninstall test
+```
+
+Uninstall the crd:
+
+```shell
+kubectl delete customresourcedefinitions.apiextensions.k8s.io clusters.mysql.radondb.io
```
Comment (runkecheng) [inline DEPLOY.md:42]: clusters.mysql.radondb.io -> clusters.mysql.radondb.com

**Comment by TCeason** [comment]
This pr only has one feature. I think it's better to contain one commit in this pr.

File: cluster/container/backup.go [hunk 43]
```diff
@@ -0,0 +1,108 @@
...
+func (c *backupSidecar) getCommand() []string {
+	return []string{"sidecar", "http"}
+}
+
+func (c *backupSidecar) getEnvVars() []corev1.EnvVar {
+	sctName := c.Spec.BackupSecretName
+
+	envs := []corev1.EnvVar{
+
```
Comment (TCeason) [inline cluster/container/backup.go:33]: Fix

File: sidecar/config.go [hunk 44]
```diff
@@ -108,7 +132,58 @@ func NewConfig() *Config {
...
+		"--host=127.0.0.1",
+		fmt.Sprintf("--user=%s", user),
+		fmt.Sprintf("--target-dir=%s", tmpdir),
+	}
+
+	return append(xtrabackupArgs, cfg.XtrabackupExtraArgs...)
+}
+
+func (cfg *Config) XCloudArgs() []string {
+
```
Comment (TCeason) [inline sidecar/config.go:174]: Fix.

**Review by andyli029 — CHANGES_REQUESTED** [review]
Keep the pr, submit another new pr.

File: config/samples/mysql_v1alpha1_cluster.yaml (hunk 17 above)
Comment (acekingke) [reply config/samples/mysql_v1alpha1_cluster.yaml:82]: ok,change to 20Gi

File: config/samples/nfs_rc.yaml (hunk 18 above)
Comment (acekingke) [reply config/samples/nfs_rc.yaml:2]: create for nfs server need it:
at first
 `kubectl apply -f nfs_rc.yaml`
then
 `kubectl apply -f nfs_server.yaml`
nfs server create success

File: Dockerfile.sidecar (hunk 4 above)
Comment (acekingke) [reply Dockerfile.sidecar:55]: restore.sh do not exist again, use golang code generate it

**Pull request closed without merging**